### DISCUSSION TITLE
Field required Error with 422 Code on Enum but not on list[Enum]

### DISCUSSION THREAD

**Discussion by PJ-Schulz** [fastapi/fastapi#11171] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from enum import Enum
from fastapi import FastAPI
import json
from fastapi.testclient import TestClient

app = FastAPI()

class MyEnum(Enum):
    foo = "foo"
    bar = "bar"
    baz = "baz"

@app.put("/abc")
def abc(value: MyEnum):
    ...

@app.put("/qwert")
def qwert(value: list[MyEnum]):
    ...

client = TestClient(app)
response = client.put('/abc', content=json.dumps("foo"))
print(response, response.text)
response = client.put('/qwert', content=json.dumps(["foo"]))
print(response, response.text)
```

### Description

I encountered an inconsistency in the behavior of FastAPI when dealing with parameters of type `Enum` and `lists of Enums`. I have two functions, abc and qwert, defined in my FastAPI application.

The abc function requires a parameter value of type MyEnum, while the qwert function requires a parameter value of type list[MyEnum]. However, when I call both functions using the same TestClient and provide the parameters using `content=` parameter, they exhibit different behaviors. This is the output of the example code:

```
<Response [422 Unprocessable Entity]> {"detail":[{"type":"missing","loc":["query","value"],"msg":"Field required","input":null,"url":"https://errors.pydantic.dev/2.6/v/missing"}]}
<Response [200 OK]> null
```

In the first put call, the error sasy that i do not pass the parameter value, so `Field required` occurs. In the second put call i do it the same, but here it works.

This inconsistency in behavior is confusing and makes it challenging to predict how FastAPI will handle Enum parameters in different scenarios. I believe this issue could potentially impact other users as well, leading to unexpected errors and behaviors in their applications.

### Operating System

Linux

### FastAPI Version

0.109.2

### Pydantic Version

2.6.0

### Python Version

3.11.7

### Additional Context

with pydantic 1.10.14 and fastapi 0.99.1 it has worked without 422 Error

**Comment by JavierSanchezCastro** [comment]
As I understand, the first example (/abc) will evaluate en MyNum as a Query parameter, while the second example (/qwert) will evaluate it as Body.

So, as I read here https://fastapi.tiangolo.com/tutorial/testing/#extended-testing-file

The new code would be:

```python3
from enum import Enum
from fastapi import FastAPI
import json
from fastapi.testclient import TestClient

app = FastAPI()

class MyEnum(Enum):
    foo = "foo"
    bar = "bar"
    baz = "baz"

@app.put("/abc")
def abc(value: MyEnum):
    return value

@app.put("/qwert")
def qwert(value: list[MyEnum]):
    return value

client = TestClient(app)
response = client.put('/abc?value=foo')
print(response, response.text)
response = client.put('/qwert', content=json.dumps(["foo"]))
print(response, response.text)
```

**Reply by PJ-Schulz** [reply]
Although the workaround by using query parameters solves the immediate problem, it does not solve the underlying inconsistency in FastAPI's handling of enum parameters via JSON content.

The inconsistency still exists between individual enum parameters and lists of enum parameters provided via JSON content. This discrepancy can lead to confusion and problems in maintaining consistent API usage patterns across different projects and applications. I've been sitting on this for a very long time, despairing as to why this should suddenly no longer work, even though it was still possible in old versions of fastapi.

**Reply by JavierSanchezCastro** [reply]
What I mean, is that the first endpoint you wrote, is going to be interpreted as query params, so makes no sense you send the data via JSON, while the second endpoint is going to be interpreted as Body, then you can send data via JSON.
Despite that, I tried with the versions you mentioned before, and it's not working for me, so I can understand it's not an error.
On the other hand, if you still have problems with TestClients, could be more related to httpx than to FastAPI.

Another way to send query parameters (from httpx doc):
```python3
#response = client.put('/abc', params={"value": ["foo"]})
response = client.put('/abc', params=QueryParams({"value": ["foo"]}))
print(response, response.text)
print(response.request.url)
```

**Reply by JavierSanchezCastro** [reply, marked as answer]
For example, in that other case:

```python3
@app.put("/abc_body")
def abc_body(value: Annotated[MyEnum, Body()]):
    return value
```

The test will be:
```python3
client = TestClient(app)
response = client.put('/abc_body', content=json.dumps("foo"))
print(response, response.text)
print(response.request.url)
```

The test is going to pass since the endpoint now is interpreted as Body

**Reply by PJ-Schulz** [reply]
For my approach the usage of `Annotated[MyEnum, Body()]` does resolve my problem because now I can call different routes always with `content=json.dumps(...)`. 

For others who come across this discussion: however, it is now no longer possible to pass the value as a normal parameter, as it always expects the value in the body.
